lighthouse-core/audits/byte-efficiency/unused-images.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+ /**
+  * @fileoverview Checks to see if images are displayed only outside of the viewport.
+  */
+'use strict';
+
+const Audit = require('./byte-efficiency-audit');
+const URL = require('../../lib/url-shim');
+
+const ALLOWABLE_OFFSCREEN_X = 100;
+const ALLOWABLE_OFFSCREEN_Y = 200;
+
+const IGNORE_THRESHOLD_IN_BYTES = 2048;
+
+class UnusedImages extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Images',
+      name: 'unused-images',
+      description: 'Unused images',
+      helpText: 'Images that are not above the fold should be lazy loaded on interaction',
+      requiredArtifacts: ['ImageUsage', 'ViewportDimensions', 'networkRecords']
+    };
+  }
+
+  /**
+   * @param {{top: number, right: number, bottom: number, left: number}} imageRect
+   * @param {{scrollWidth: number, scrollHeight: number}} viewportDimensions
+   * @return {number}
+   */
+  static computeVisiblePixels(imageRect, viewportDimensions) {
+    const scrollWidth = viewportDimensions.scrollWidth;
+    const scrollHeight = viewportDimensions.scrollHeight;
+
+    const top = Math.max(imageRect.top, -1 * ALLOWABLE_OFFSCREEN_Y);
+    const right = Math.min(imageRect.right, scrollWidth + ALLOWABLE_OFFSCREEN_X);
+    const bottom = Math.min(imageRect.bottom, scrollHeight + ALLOWABLE_OFFSCREEN_Y);
+    const left = Math.max(imageRect.left, -1 * ALLOWABLE_OFFSCREEN_X);
+
+    return Math.max(right - left, 0) * Math.max(bottom - top, 0);
+  }
+
+  /**
+   * @param {!Object} image
+   * @param {{scrollWidth: number, scrollHeight: number}} viewportDimensions
+   * @return {?Object}
+   */
+  static computeWaste(image, viewportDimensions) {
+    const url = URL.getDisplayName(image.src, {preserveQuery: true});
+    const totalPixels = image.clientWidth * image.clientHeight;
+    const visiblePixels = this.computeVisiblePixels(image.clientRect, viewportDimensions);
+    const wastedRatio = 1 - visiblePixels / totalPixels;
```

Comment:
True, I open to one that would do set subtraction but that seemed too specific to this audit to put in image usage and felt weird breaking out into a new one. It also meant the settings for failure (margin and so forth) were located in the gatherer rather than with the audit.

We have a spectrum of this throughout LH though, so I'm curious what other opinions are. I'm generally more in favor of keeping gatherers purely informational/more logic-less and having failure conditions measured in audits.